### PR TITLE
integration tests for active learning

### LINKED ISSUE #439: Active learning acquisition functions are missing integration tests

### PR DESCRIPTION
**Pull request by hstojic**
active learning algos were missing integration tests, this would close #439

tests are a bit slow, which is to be expected I guess, since we want to be reasonably sure that we are really learning the function or the feasibility set we do need to run more iterations than usual

something to discuss: should we cover batch versions, async rules, trust region, tensorboard? (mirroring the optimization tests...)

### CODE REVIEW

File: tests/integration/test_active_learning.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        BayesianOptimizer(observer, search_space)
+        .optimize(num_steps, initial_data, model, acquisition_rule)
+        .try_get_final_model()
+    )
+
+    test_query_points = search_space.sample_sobol(10000 * search_space.dimension)
+    test_data = observer(test_query_points)
+    predicted_means, _ = final_model.model.predict_f(test_query_points)  # type: ignore
+
+    npt.assert_allclose(predicted_means, test_data.observations, rtol=0.5)
```
Comment (vpicheny) [inline tests/integration/test_active_learning.py]: just for checking: allclose would still use the default atol, which is quite restrictive, right? otherwise, this is very loose?

File: tests/integration/test_active_learning.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        (150, EfficientGlobalOptimization(ExpectedFeasibility(80, delta=1)), 80),
+        (150, EfficientGlobalOptimization(ExpectedFeasibility(80, delta=2)), 80),
+        (150, EfficientGlobalOptimization(ExpectedFeasibility(20, delta=1)), 20),
+    ],
+)
+def test_optimizer_learns_feasibility_set_of_thresholded_branin_function(
+    num_steps: int, acquisition_rule: AcquisitionRule[TensorType, SearchSpace], threshold: int
+) -> None:
+    """
+    Ensure that the objective function is effecitvely learned, such that the final model
```
Comment (vpicheny) [inline tests/integration/test_active_learning.py]: typo

File: tests/integration/test_active_learning.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        )
+        kernel.lengthscales.prior = tfp.distributions.LogNormal(
+            tf.math.log(kernel.lengthscales), prior_scale
+        )
+        gpr = gpflow.models.GPR(data.astuple(), kernel, noise_variance=1e-5)
+        gpflow.set_trainable(gpr.likelihood, False)
+
+        return GaussianProcessRegression(gpr)
+
+    num_initial_points = 20
```
Comment (vpicheny) [inline tests/integration/test_active_learning.py]: can we start with fewer? 20 points is already a lot, so it's possible that the `PredictiveVariance` doesn't do much.

Or we could check if the error decreases? this might be a bit tricky though...

File: tests/integration/test_active_learning.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
```
Comment (vpicheny) [inline tests/integration/test_active_learning.py]: same as before: 20 seems really large. Try 6?

File: tests/integration/test_active_learning.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    model = build_model(initial_data)
+    final_model = (
+        BayesianOptimizer(observer, search_space)
+        .optimize(num_steps, initial_data, model, acquisition_rule)
+        .try_get_final_model()
+    )
+
+    test_points = search_space.sample_sobol(10000 * search_space.dimension)
+    prob = _excursion_probability(test_points, final_model, threshold)
+    accuracy = tf.reduce_sum(prob * (1 - prob))
```
Comment (vpicheny) [inline tests/integration/test_active_learning.py]: use mean instead of sum? so it doesn't depend on the number of test points

File: tests/integration/test_active_learning.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+    assert accuracy < 0.0001
```
Comment (vpicheny) [inline tests/integration/test_active_learning.py]: this looks like an incredibly hard target to reach

File: tests/integration/test_active_learning.py (hunk 1 above)
Comment (vpicheny) [reply tests/integration/test_active_learning.py]: Also, this is not really consistent with the other test: the other tests a single average measure of accuracy, while this is a max. Max is fine, but maybe rewrite `crit = np.abs(predicted_means, test_data.observations)` and assert crit < something?

Comment (hstojic) [reply tests/integration/test_active_learning.py]: yes, accuracy criteria is something to discuss, its difficult to say what to expect in terms of global accuracy with 100 observations, and we dont want to have tests that take ages - I'm open to suggestions here

the other test - you mean feasibility set? thats a different type of problem, but I guess we could take a mean here and be happy if its smth small?

File: tests/integration/test_active_learning.py (hunk 3 above)
Comment (hstojic) [reply tests/integration/test_active_learning.py]: sure, we can try starting with fewer points

I thought about doing the decrease in error test here but concluded that would be more suitable as a unit test

File: tests/integration/test_active_learning.py (hunk 4 above)
Comment (hstojic) [reply tests/integration/test_active_learning.py]: ok, will do

File: tests/integration/test_active_learning.py (hunk 6 above)
Comment (hstojic) [reply tests/integration/test_active_learning.py]: yeah, it seemed like it works but doesn't at the end, 1 works

File: tests/integration/test_active_learning.py (hunk 1 above)
Comment (vpicheny) [reply tests/integration/test_active_learning.py]: For this test, we probably want a criterion that would fail with `num_initial_points` and pass with more points, provided that they are not pathologically distributed (eg they should be roughly space-filling, not concentrated in one corner of the space). We can't do much more than that (eg we could see if we beat random search over replicated experiments, but we won't do that). But this is already a good test.

Comment (vpicheny) [reply tests/integration/test_active_learning.py]: Max error is probably better than average error because it will reward more space-filling designs.

File: tests/integration/test_active_learning.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    test_range = (tf.reduce_max(test_data.observations) - tf.reduce_min(test_data.observations))
+    criterion = 0.01 * test_range
+
+    # we expect a model with initial data to fail the criterion
+    initial_model = build_model(initial_data)
+    initial_model.optimize(initial_data)
+    initial_predicted_means, _ = initial_model.model.predict_f(test_query_points)  # type: ignore
+    initial_accuracy = tf.reduce_max(tf.abs(initial_predicted_means - test_data.observations))
+
+    assert not initial_accuracy < criterion
```
Comment (hstojic) [inline tests/integration/test_active_learning.py:90]: we test now that initial model fails and final model improves on this and satisfies the criterion

File: tests/integration/test_active_learning.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    initial_query_points = search_space.sample_halton(num_initial_points)
+    observer = mk_observer(scaled_branin)
+    initial_data = observer(initial_query_points)
+
+    # we set a performance criterion at 1% of the range
+    # max absolute error needs to be bettter than this criterion
+    test_query_points = search_space.sample_sobol(10000 * search_space.dimension)
+    test_data = observer(test_query_points)
+    test_range = (tf.reduce_max(test_data.observations) - tf.reduce_min(test_data.observations))
+    criterion = 0.01 * test_range
```
Comment (hstojic) [inline tests/integration/test_active_learning.py:82]: max abs error shouldn't cross 1% of the range, that seems like a good global precision to me?

File: tests/integration/test_active_learning.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+    num_initial_points = 6
+    initial_query_points = search_space.sample_halton(num_initial_points)
+    observer = mk_observer(branin)
+    initial_data = observer(initial_query_points)
+
+    # we set a performance criterion at 0.001 probability of required precision per point
+    n_test = 10000 * search_space.dimension
+    test_query_points = search_space.sample_sobol(n_test)
+    criterion = 0.001 * (1 - 0.001) * tf.cast(n_test, tf.float64)
```
Comment (hstojic) [inline tests/integration/test_active_learning.py]: new criterion, 0.001/0.999 predicted probability per tested point seems like a reasonable accuracy to me, any thoughts?

File: tests/integration/test_active_learning.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        return GaussianProcessRegression(gpr)
+
+    num_initial_points = 6
+    initial_query_points = search_space.sample_halton(num_initial_points)
+    observer = mk_observer(branin)
+    initial_data = observer(initial_query_points)
+
+    # we set a performance criterion at 0.001 probability of required precision per point
+    n_test = 10000 * search_space.dimension
+    test_query_points = search_space.sample_sobol(n_test)
```
Comment (hstojic) [inline tests/integration/test_active_learning.py]: I think we are more interested in model accuracy around the threshold (or perhaps the whole feasible region?), instead of getting test points from the whole search space, we should then collect test points around the boundary - we could do something like:
```
test_query_points = search_space.sample(100000)
test_data = observer(test_query_points)

mask = tf.reduce_all(tf.concat([test_data.observations > 80 - 5, test_data.observations < 80 + 5], axis = 1), axis=1)
threshold_test_points = tf.boolean_mask(test_query_points, mask)
assert threshold_test_points.shape[0] > 1000
```

Comment (vpicheny) [reply tests/integration/test_active_learning.py]: I'm not sure I 100% agree, we also want our GP to be good enough to predict anywhere which side it's on with high probability. I think the random sample is fine.

We could also have 2 criteria, one 'global' as it is and one 'local' as you describe and we could check both. Not sure it's worth the extra effort

File: tests/integration/test_active_learning.py (hunk 9 above)
Comment (vpicheny) [reply tests/integration/test_active_learning.py]: I like this, it's principled now

**Review by vpicheny — APPROVED** [review]
great changes, looks good now!

File: tests/integration/test_active_learning.py (hunk 9 above)
Comment (hstojic) [reply tests/integration/test_active_learning.py]: hm, its sensitive to the volume of the search space occupied by the boundary, for the threshold of 20 boundary is much larger and then there are more points with smaller accuracy, so the 3rd test actually fails here :|

Comment (hstojic) [reply tests/integration/test_active_learning.py]: perhaps a quick monte carlo estimate of the volume of the feasible space and make the criterion dependent on that as well?

Comment (hstojic) [reply tests/integration/test_active_learning.py]: although its reasonable to expect to need to have more observations to achieve the same accuracy for a larger volume, so we could also simply increase number of steps for threshold 20

File: tests/integration/test_active_learning.py (hunk 10 above)
Comment (hstojic) [reply tests/integration/test_active_learning.py]: I went with the 2 criteria version...
